mmf/models/mmf_transformer.py

```diff
@@ -193,133 +57,111 @@ def build_encoders(self):
             for param in self.image_encoder.parameters():
                 param.requires_grad = False
 
-    def build_embeddings(self):
-        """Initialize the embedding class we will use for multiple
-        modalities (here just text and image). For the text embeeddings we will use the
-        pretrained weights from the trasnformer model rather than training from scratch.
-        """
-        self.embeddings = MMFTransformerEmbeddings(
-            self.config, self.transformer_config, self.transformer
-        )
-
     def build_heads(self):
         """Initialize the classifier head. It takes the output of the
         transformer encoder and passes it through a pooler (we use the pooler from BERT
         model), then dropout, BertPredictionHeadTransform (which is a linear layer,
         followed by activation and layer norm) and lastly a linear layer projecting the
         hidden output to classification labels.
         """
+        transformer_config = self.backend.get_config()
```

Comment:
So, here's a thought, I don't know specifics of pytext but I think something like this would be better:

- Initialize the backend in init
- backend has `get_pooler` function similar to get_config() function, then you don't need to pass config to it specifically and it can stay backend agnostic
- Similarly, config values can become properties of backend class which will be agnostic again to backend.
- Same for prediction head transform.
- This can be extended to other BERT module layers as well. 
- Now backend also handle cases which are not regular like difference in xlm-r and bert etc in the hidden shadows itself. 

If something is missing in a model, backend can then return identity class.

Reply:
I think for these layers in the `head` it's fine to get the config and build it in the main model itself. These are not BERT specific(although they are taken from bert model). Additionally all the different backends don't provide any specific layers for classifiers or poolers. So for those who don't provide we will require calling a different backend(which I think is cumbersome). They make these changes in the model forward pass itself. So in order to keep the forward return type same for all backends, the pooler or classifier layers should  be built in the main model IMO.